### PR TITLE
upload-aws-s3: Add file url building function

### PR DESCRIPTION
**Pull request by jhoward1994**
### What does it do?

Adds optional config to the aws s3 provider so users can build file urls with custom logic

### Why is it needed?

To solve needs such as https://github.com/strapi/strapi/pull/17495 without adding more edge cases to the if else statement

### How to test it?

Added a unit test
Set up and use the aws s3 provider

### CODE REVIEW

File: packages/providers/upload-aws-s3/README.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ npm install @strapi/provider-upload-aws-s3 --save
   - `signedUrlExpires` is the number of seconds before a signed URL expires. (See [how signed URLs work](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-urls.html)). Defaults to 15 minutes and URLs are only signed when ACL is set to `private`.
   - `Bucket` is the name of the bucket to upload to.
 - `actionOptions` is passed directly to the parameters to each method respectively. You can find the complete list of [upload/ uploadStream options](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property) and [delete options](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObject-property)
+- `buildFileUrl` optionally pass this function to define custom logic for building file url. This is provided as a way to handle the many different s3 provider options.
```
Comment (Marc-Roig) [inline packages/providers/upload-aws-s3/README.md:33]: I was thinking if this name is expressive enough to understand what  is doing.

At first it might mean multiple things,
- Build url before being uploaded
- Build url for url signing
- Build url after it has been upload

wdyt if we find alternative names that make this difference more clear? A name that conveys it decodes the uploaded file data into the url that will be saved into db?

Comment (jhoward1994) [reply packages/providers/upload-aws-s3/README.md:33]: Yeah good point. How about `mapFileDataToDbUrl` ?

Comment (Marc-Roig) [reply packages/providers/upload-aws-s3/README.md:33]: That definetly sounds bettter, I guess `mapFileDataToUrl`  without "Db" would be expressive enough!

I wonder if adding something like:
```
config: {
   ...
   mappers: {
     fileDataToUrl: () => {}
    }
 }
 ```
 
 would be more extensible or if its just too much

File: packages/providers/upload-aws-s3/README.md [hunk 2]
```diff
@@ -30,6 +30,7 @@ npm install @strapi/provider-upload-aws-s3 --save
   - `signedUrlExpires` is the number of seconds before a signed URL expires. (See [how signed URLs work](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-urls.html)). Defaults to 15 minutes and URLs are only signed when ACL is set to `private`.
   - `Bucket` is the name of the bucket to upload to.
 - `actionOptions` is passed directly to the parameters to each method respectively. You can find the complete list of [upload/ uploadStream options](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property) and [delete options](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObject-property)
+- `mapFileDataToUrl` optionally pass this function to define custom logic for building file url. This is provided as a way to handle the many different s3 provider options.
```
Comment (Marc-Roig) [inline packages/providers/upload-aws-s3/README.md:33]:
```suggestion
- `mapFileDataToUrl` optionally pass this function to define custom logic for building file url from the uploaded file data. This will be the url that will be stored in the database. This is provided as a way to handle the many different s3 provider options.
```

**Comment by hanpaine** [comment]
Hey @jhoward1994 did this PR need to go anywhere? Does it need much to resolve or needs closing/abandoning?

**Comment by jhoward1994** [comment]
Wow yes sorry it has been a while on this one. It still looks like a good idea to me however the code for this provider on `develop` has changed significantly.

It's not a huge change so shouldn't be too hard to make it work with develop. I can take a look when I have time between v5 stuff, is that okay?

**Comment by alexandrebodin** [comment]
> [!IMPORTANT] 
> With the release of Strapi 5, we have lifted the code freeze. To ensure we can effectively review and assist with contributions, we are now only accepting PRs targeting version 5 (the develop branch).<br/> 
> As your PR targets version 4, it has been automatically closed. However, we would greatly appreciate it if you could update your contribution to align with version 5 and resubmit!
